File: anyharness/sdk/src/client/git.ts
import type {
  CommitRequest,
  CommitResponse,
  GitBranchDiffFilesResponse,
  GitBranchRef,
  GitDiffOptions,
  GitDiffResponse,
  GitStatusSnapshot,
  ListBranchDiffFilesOptions,
  PushRequest,
  PushResponse,
  RenameBranchRequest,
  RenameBranchResponse,
  StagePathsRequest,
  UnstagePathsRequest,
} from "../types/git.js";
import { withTimingCategory, type AnyHarnessRequestOptions, type AnyHarnessTransport } from "./core.js";

type GitDiffClientOptions = GitDiffOptions & {
  request?: AnyHarnessRequestOptions;
};

type ListBranchDiffFilesClientOptions = ListBranchDiffFilesOptions & {
  request?: AnyHarnessRequestOptions;
};

export class GitClient {
  constructor(private readonly transport: AnyHarnessTransport) {}

  async getStatus(
    workspaceId: string,
    options?: AnyHarnessRequestOptions,
  ): Promise<GitStatusSnapshot> {
    return this.transport.get<GitStatusSnapshot>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/status`,
      withTimingCategory(options, "git.status"),
    );
  }

  async getDiff(
    workspaceId: string,
    path: string,
    options: GitDiffClientOptions = {},
  ): Promise<GitDiffResponse> {
    const params: Array<[string, string]> = [["path", path]];
    const scope = options.scope ?? null;
    if (scope && scope !== "working_tree") {
      params.push(["scope", scope]);
    }
    if (scope === "branch") {
      const baseRef = options.baseRef?.trim();
      const oldPath = options.oldPath?.trim();
      if (baseRef) {
        params.push(["baseRef", baseRef]);
      }
      if (oldPath) {
        params.push(["oldPath", oldPath]);
      }
    }

    return this.transport.get<GitDiffResponse>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/diff?${encodeQueryParams(params)}`,
      withTimingCategory(options.request, "git.diff"),
    );
  }

  async listBranchDiffFiles(
    workspaceId: string,
    options: ListBranchDiffFilesClientOptions = {},
  ): Promise<GitBranchDiffFilesResponse> {
    const params: Array<[string, string]> = [];
    const baseRef = options.baseRef?.trim();
    if (baseRef) {
      params.push(["baseRef", baseRef]);
    }
    const query = encodeQueryParams(params);
    return this.transport.get<GitBranchDiffFilesResponse>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/diff/branch-files${query ? `?${query}` : ""}`,
      withTimingCategory(options.request, "git.branch_diff_files"),
    );
  }

  async listBranches(
    workspaceId: string,
    options?: AnyHarnessRequestOptions,
  ): Promise<GitBranchRef[]> {
    return this.transport.get<GitBranchRef[]>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/branches`,
      options,
    );
  }

  async renameBranch(workspaceId: string, newName: string): Promise<RenameBranchResponse> {
    return this.transport.post<RenameBranchResponse>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/rename-branch`,
      { newName } satisfies RenameBranchRequest,
    );
  }

  async stagePaths(workspaceId: string, paths: string[]): Promise<void> {
    await this.transport.post<void>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/stage`,
      { paths } satisfies StagePathsRequest,
    );
  }

  async unstagePaths(workspaceId: string, paths: string[]): Promise<void> {
    await this.transport.post<void>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/unstage`,
      { paths } satisfies UnstagePathsRequest,
    );
  }

  async commit(workspaceId: string, input: CommitRequest): Promise<CommitResponse> {
    return this.transport.post<CommitResponse>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/commit`,
      input,
    );
  }

  async push(workspaceId: string, input: PushRequest = {}): Promise<PushResponse> {
    return this.transport.post<PushResponse>(
      `/v1/workspaces/${encodeURIComponent(workspaceId)}/git/push`,
      input,
    );
  }
}

function encodeQueryParams(params: Array<[string, string]>): string {
  return params
    .map(([key, value]) => `${key}=${encodeURIComponent(value)}`)
    .join("&");
}

File: anyharness/crates/anyharness-contract/src/v1/git.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// ---------------------------------------------------------------------------
// Status snapshot — the main UI contract for workspace git state
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitStatusSnapshot {
    pub workspace_id: String,
    pub workspace_path: String,
    pub repo_root_path: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_branch: Option<String>,
    pub head_oid: String,
    pub detached: bool,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub upstream_branch: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub suggested_base_branch: Option<String>,
    pub ahead: u32,
    pub behind: u32,

    pub operation: GitOperation,
    pub conflicted: bool,
    pub clean: bool,

    pub summary: GitStatusSummary,
    pub actions: GitActionAvailability,
    pub files: Vec<GitChangedFile>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum GitOperation {
    #[default]
    None,
    Merge,
    Rebase,
    CherryPick,
    Revert,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitStatusSummary {
    pub changed_files: u32,
    pub additions: u32,
    pub deletions: u32,
    pub included_files: u32,
    pub conflicted_files: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitActionAvailability {
    pub can_commit: bool,
    pub can_push: bool,
    pub push_label: String,
    pub can_create_pull_request: bool,
    pub can_create_draft_pull_request: bool,
    pub can_create_branch_workspace: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason_if_blocked: Option<String>,
}

// ---------------------------------------------------------------------------
// Changed file entry
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitChangedFile {
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub old_path: Option<String>,
    pub status: GitFileStatus,
    pub additions: u32,
    pub deletions: u32,
    pub binary: bool,
    pub included_state: GitIncludedState,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum GitFileStatus {
    Modified,
    Added,
    Deleted,
    Renamed,
    Copied,
    Untracked,
    Conflicted,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum GitIncludedState {
    Included,
    Excluded,
    Partial,
}

// ---------------------------------------------------------------------------
// Diff
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum GitDiffScope {
    WorkingTree,
    Unstaged,
    Staged,
    Branch,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitDiffRequest {
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitDiffResponse {
    pub path: String,
    pub scope: GitDiffScope,
    pub binary: bool,
    pub truncated: bool,
    pub additions: u32,
    pub deletions: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolved_base_oid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub merge_base_oid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub head_oid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub patch: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitDiffFile {
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub old_path: Option<String>,
    pub status: GitFileStatus,
    pub additions: u32,
    pub deletions: u32,
    pub binary: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchDiffFilesResponse {
    pub base_ref: String,
    pub resolved_base_oid: String,
    pub merge_base_oid: String,
    pub head_oid: String,
    pub files: Vec<GitDiffFile>,
}

// ---------------------------------------------------------------------------
// Branches
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchRef {
    pub name: String,
    pub is_remote: bool,
    pub is_head: bool,
    pub is_default: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub upstream: Option<String>,
}

// ---------------------------------------------------------------------------
// Rename branch
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct RenameBranchRequest {
    pub new_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct RenameBranchResponse {
    pub old_name: String,
    pub new_name: String,
}

// ---------------------------------------------------------------------------
// Stage / unstage
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct StagePathsRequest {
    pub paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct UnstagePathsRequest {
    pub paths: Vec<String>,
}

// ---------------------------------------------------------------------------
// Commit
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct CommitRequest {
    pub summary: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct CommitResponse {
    pub oid: String,
    pub summary: String,
}

// ---------------------------------------------------------------------------
// Push
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PushRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub remote: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PushResponse {
    pub remote: String,
    pub branch: String,
    pub published: bool,
}
